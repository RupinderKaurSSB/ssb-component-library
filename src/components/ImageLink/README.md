ImageLink
========

> Clickable image box with title

### Usage

#### HTML
```html
<a class="ssb-image-link" href=" ">
    <div class="image-background"><img src="" alt="" /></div>
    <span class="il-type">Type</span>
    <span class="il-title">Title</span>
    {feather.arrowRightCircle 32px}
</a>
```

#### React
```jsx harmony
<ImageLink image={<img src={doge} alt="image" />} onClick={function} type="Type" title="Tittel" />
```
```jsx harmony
<ImageLink image={<img src={doge} alt="image" />} link="./" type="Type" title="Tittel" />
```
Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| className   | string | Optional container class|
| image | required element | An image to be put as background |
| link | string | Link to be navigated to |
| onClick | func | Function callback to be used instead of link |
| title | string | Title text |
| type | string | Type text |
