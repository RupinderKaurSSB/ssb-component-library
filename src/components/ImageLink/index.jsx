import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { ArrowRight, ArrowRightCircle } from 'react-feather';

export function useHover() {
	const [value, setValue] = useState(false);

	const hoverRef = useRef(null);

	useEffect(
		// eslint-disable-next-line consistent-return
		() => {
			const handleMouseOver = () => setValue(true);
			const handleMouseOut = () => setValue(false);
			const element = hoverRef && hoverRef.current;

			if (element) {
				element.addEventListener('mouseover', handleMouseOver);
				element.addEventListener('mouseout', handleMouseOut);
				return () => {
					element.removeEventListener('mouseover', handleMouseOver);
					element.removeEventListener('mouseout', handleMouseOut);
				};
			}
		},
		[hoverRef],
	);

	return [hoverRef, value];
}


const ImageLink = ({ className, image, link, onClick, orientation, title, type }) => {
	const [hoverRef, hovered] = useHover();
	return (
		<a className={`ssb-image-link ${orientation}${className ? ` ${className}` : ''}`} href={link} onClick={onClick} ref={hoverRef}>
			<div className="image-background">{image}</div>
			<span className="il-type">{type}</span>
			<span className="il-title">{title}</span>
			{hovered ? <ArrowRightCircle className="arrow-icon" size={32} /> : <ArrowRight className="arrow-icon" size={32} /> }
		</a>
	);
};

ImageLink.defaultProps = {
	onClick: () => {},
	orientation: 'vertical',
};

ImageLink.propTypes = {
	className: PropTypes.string,
	image: PropTypes.element.isRequired,
	link: PropTypes.string,
	onClick: PropTypes.func,
	orientation: PropTypes.oneOf(['horizontal', 'vertical']),
	title: PropTypes.string,
	type: PropTypes.string,
};

export default ImageLink;
