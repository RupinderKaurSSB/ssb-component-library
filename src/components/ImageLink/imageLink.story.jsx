import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import ImageLink from './index';
import doge from './doge.png';
import jokeDog from './jokeDog.png';

const sup = (() => {
	console.log('Image link clicked');
});

storiesOf('Image Link', module).addDecorator(centered)
	.add('Default', () => (
		<ImageLink image={<img src={jokeDog} alt="" />} onClick={sup} type="Type" title="Tittel" />
	))
	.add('Horizontal', () => (
		<ImageLink orientation="horizontal" image={<img src={doge} alt="" />} onClick={sup} type="Type" title="Tittel" />
	));
