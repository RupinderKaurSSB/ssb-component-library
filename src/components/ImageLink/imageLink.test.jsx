import React from 'react';
import { shallow } from 'enzyme';
import ImageLink from './index';
import doge from './doge.png';

describe('ImageLink component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<ImageLink image={<img src={doge} alt="image" />} onClick={() => {}} type="Type" title="Tittel" />);
		expect(wrapper).toMatchSnapshot();
	});
	test('Sets default orientation className', () => {
		const wrapper = shallow(<ImageLink image={<img src={doge} alt="image" />} onClick={() => {}} type="Type" title="Tittel" />);
		expect(wrapper.hasClass('vertical')).toEqual(true);
	});
	test('Switches className on orientation change', () => {
		const wrapper = shallow(<ImageLink orientation="horizontal" image={<img src={doge} alt="image" />} onClick={() => {}} type="Type" title="Tittel" />);
		expect(wrapper.hasClass('horizontal')).toEqual(true);
	});
});
