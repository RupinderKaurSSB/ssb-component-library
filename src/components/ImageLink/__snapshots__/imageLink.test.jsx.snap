// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLink component Matches the snapshot 1`] = `
<a
  className="ssb-image-link vertical"
  onClick={[Function]}
>
  <div
    className="image-background"
  >
    <img
      alt="image"
      src=""
    />
  </div>
  <span
    className="il-type"
  >
    Type
  </span>
  <span
    className="il-title"
  >
    Tittel
  </span>
  <ArrowRightCircle
    color="currentColor"
    size={32}
  />
</a>
`;
